### PR TITLE
Add API query for available per-process CUDA memory

### PR DESCRIPTION
**Pull request by benjaminglass1**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141699
* #141371
* #141370
* #141369
* #141580
* #141368
* __->__ #140620
* #141367

Certain `cpp_wrapper`-enabled tests were OOM-ing in the CI pipeline, with error messages suggesting that sufficient memory was accessible.  This ultimately resulted from an internal memory limitation that was not queryable in the API.  This PR adds querying for that limit.

Additionally, the failing tests had incorrect memory availability checks, and are updated with measured memory requirements.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Review by malfet — APPROVED** [review]
Sure, but I wonder if we can have a better API for everyone, not just for our unit tests (i.e.  one should be able to query memory fraction allocated to the current process

File: torch/testing/_internal/common_device_type.py [hunk 1]
```diff
@@ -1312,7 +1314,13 @@ def _has_sufficient_memory(device, size):
         # torch.cuda.mem_get_info, aka cudaMemGetInfo, returns a tuple of (free memory, total memory) of a GPU
         if device == "cuda":
             device = "cuda:0"
-        return torch.cuda.memory.mem_get_info(device)[0] >= size
+        return (
+            min(
+                MAX_CUDA_MEM if MAX_CUDA_MEM else inf,
```
[the remark this answers is not in the document]
Comment (benjaminglass1) [reply torch/testing/_internal/common_device_type.py]: No longer used at all.

File: torch/testing/_internal/common_device_type.py [hunk 2]
```diff
@@ -1312,7 +1314,13 @@ def _has_sufficient_memory(device, size):
...
+                torch.cuda.memory.mem_get_info(device)[0],
```
[the remark this answers is not in the document]
Comment (benjaminglass1) [reply torch/testing/_internal/common_device_type.py]: @malfet New API pushed; please re-review.

File: torch/testing/_internal/common_device_type.py [hunk 3]
```diff
@@ -1312,7 +1313,7 @@ def _has_sufficient_memory(device, size):
         # torch.cuda.mem_get_info, aka cudaMemGetInfo, returns a tuple of (free memory, total memory) of a GPU
         if device == "cuda":
             device = "cuda:0"
-        return torch.cuda.memory.mem_get_info(device)[0] >= size
+        return min(MAX_CUDA_MEM, torch.cuda.memory.mem_get_info(device)[0]) >= size
```
[the remark this answers is not in the document]
Comment (benjaminglass1) [reply torch/testing/_internal/common_device_type.py]: @amjames Does the new API returning the memory fraction address your concern?

**Comment by benjaminglass1** [comment]
Note for review: there's an identical, separate API for MPS. I chose not to add that yet due to not needing it, but am willing to do so if there's a preference for it.

**Review by eqy — APPROVED** [review]
thanks for this change! looking forward to using it to fix some CUDA tests

**Pull request closed without merging**